src/box/xlog_reader.c

```diff
@@ -0,0 +1,518 @@
+#include <stdint.h>
+#include <small/ibuf.h>
+#include <small/lsregion.h>
+
+#include "xlog_reader.h"
+
+#include "cbus.h"
+#include "fiber.h"
+#include "fiber_cond.h"
+#include "memory.h"
+#include "tt_static.h"
+#include "xlog.h"
+
+/**
+ * Stream message. It is used to pass batches with xlog requests from reader
+ * thread to TX and back to be freed.
+ */
+struct xlog_reader_stream_msg {
+	/** Base class. */
+	struct cmsg base;
+	/** Xlog reader instance. */
+	struct xlog_reader *reader;
+	/**
+	 * Requests of the batch.
+	 * Request is object of `struct recovery_request`.
+	 */
+	struct stailq requests;
+	/** Position of the batch in reader batches ring buffer. */
+	unsigned pos;
+	/** Error reading xlog. If it is set requests can be not empty. */
+	struct error *error;
+	/**
+	 * Set if xlog EOF is reached. The `error` can be set as well
+	 * in this case.
+	 */
+	bool eof;
+};
+
+/**
+ * Control message. It is used to send stop or finished message. See also
+ * xlog_reader::stop_msg and xlog_reader::finished_msg.
+ */
+struct xlog_reader_control_msg {
+	/** Base class. */
+	struct cmsg base;
+
+	/** Xlog reader instance. */
+	struct xlog_reader *reader;
+};
+
+/** Number xlog_reader_stream_msg messages in ring buffer. See xlog_reader. */
+#define MSG_COUNT 2
```

Comment:
Since there'll never be more than two messages (one in process, one being read), I think it's better to use a boolean flag instead.